File: solution/sort/mergesort.go
package sort

import (
	"cmp"
	"fmt"
)

func merge[T cmp.Ordered](arr []T, low int, mid int, high int) {
    left_len  := mid - low  + 1
    right_len := high - mid

    left := make([]T,  left_len,  left_len)
    right := make([]T, right_len, right_len)

    for i := 0; i < left_len; i++ {
        left[i] = arr[low + i]
    }

    for i := 0; i < right_len; i++ {
        right[i] = arr[mid + i + 1]
    }
    fmt.Println(low, mid, high)

    i, j, k := 0, 0, low
    for ; i < left_len && j < right_len; k++ {
        if left[i] > right[j] {
            arr[k] = right[j]
            j++
        } else {
            arr[k] = left[i]
            i++
        }
    }
    for ; i < left_len; i++ {
        arr[k] = left[i]
        k++
    }
    for ; j < right_len; j++ {
        arr[k] = right[j]
        k++
    }
}

func mergesort[T cmp.Ordered](arr []T, low int, high int) {
    if low >= high { return }

    mid := (low + high) / 2
    
    mergesort(arr, low,      mid)
    mergesort(arr, mid + 1, high)
    
    merge(arr, low, mid, high)
}

func RMergesort[T cmp.Ordered](arr []T) {
    mergesort(arr, 0, len(arr) - 1)
}

File: src/day1/MergeSort.ts
function merge(arr: number[], workingBuffer: number[], low: number, mid: number, high: number) {
    let i = low, j = mid + 1, k = low;
    for ( ; i <= mid && j <= high; k++) {
        if (workingBuffer[i] < workingBuffer[j]) {
            arr[k] = workingBuffer[i];
            i++;
        } else {
            arr[k] = workingBuffer[j];
            j++;
        }
    }
    for ( ; i <= mid; i++, k++) arr[k] = workingBuffer[i]; 

    for ( ; j <= high; j++, k++) arr[k] = workingBuffer[j];

    for (k = low; k <= high; k++) workingBuffer[k] = arr[k];
}

function mergeSort(arr: number[], workingBuffer: number[], low: number, high: number) {
    if (low >= high) return;

    const mid = Math.floor((low + high) / 2);

    mergeSort(arr, workingBuffer, low, mid);
    mergeSort(arr, workingBuffer, mid + 1, high);

    merge(arr, workingBuffer, low, mid, high);
}

export default function merge_sort(arr: number[]): void {
    const workingBuffer = [];
    for (let i = 0; i < arr.length; i++) workingBuffer[i] = arr[i];
    mergeSort(arr, workingBuffer, 0, arr.length - 1);
}
